### ISSUE TITLE
Proposal: a general `JointGaussian` — Gaussian conditioning as affine-map pushforward

### ISSUE THREAD

**Issue by arob5** [TARPS-group/prob-pipe#297]
> **Companion proposal.** **#299** (ensemble Kalman updates — approximate Gaussian inference via fit-then-condition) builds directly on the `AffineGaussianMap` primitive defined here.

## Summary

This proposal makes `JointGaussian` the single, general home for **all exact Gaussian
conditioning** in ProbPipe. A `JointGaussian` is a Gaussian over a structured, named event
(an `EventTemplate` of numeric blocks) carrying full cross-covariance between blocks, so any
block can be conditioned on any other.

The central design move: **Gaussian conditioning is the pushforward of the joint through an
affine map.** Conditioning constructs an `AffineGaussianMap` (the Kalman/Matheron map) and
pushes the prior through it. Push the *moments* and you get the conditional `JointGaussian`
(the parametric output); push *samples* and you get an empirical conditional (the Matheron /
ensemble-Kalman output). One computation, two outputs — and the `condition_on` op contract
("returns a `Distribution`") is preserved, because the pushforward of a Gaussian through an
affine map is closed-form and dispatches back to a `JointGaussian`.

The first implementation is **dense**, but the covariance is reached through a thin
representation seam designed so a `LinOp`-backed (structured / low-rank / canonical) form drops
in later without touching the conditioning algorithm or the public API.

Builds on / coordinates with: **#235** (`EventTemplate`), **#284** (composition algebra `*`;
`condition_on` contract), **#255** (kernels / `condition_on` dispatch), **#26** + **#67**
(`TransportMap` / pushforward registry, where `AffineGaussianMap` lives), **#31** (`LinOp`
improvements, for the future structured backing).

---

## Motivation — what exists today, and the gaps

Today's [`JointGaussian`](probpipe/distributions/_joint_gaussian.py):

- N named components ✓, but **flat-only** (no nested blocks; conditioning hand-rolls flat
  index slices rather than using the `EventTemplate`).
- **Dense** mean vector + `(d, d)` covariance; not `LinOp`-backed.
- **One conditioning formula** — parameter/observed-space `solve(Σ_oo, …)` — and it returns a
  `JointGaussian` only. No way to obtain the underlying map, no measurement-noise term, no
  choice of computational form.
- No construction from a generative (linear-Gaussian) structure or from empirical moments.

Adjacent gaps this unlocks:

- [`JointEmpirical._condition_on`](probpipe/distributions/_joint_empirical.py) **ignores the
  observed value and just drops blocks** — that is *marginalization*, mislabeled as
  conditioning. There is no value-aware empirical conditioning anywhere; the affine-map
  primitive here is what the ensemble-Kalman issue uses to provide it.
- [`GaussianRandomFunction`](probpipe/distributions/gaussian_random_function.py) has **no
  conditioning** — GP regression is Gaussian conditioning and could share this core.
- A linear-Gaussian model (`likelihood * prior`, #284) has no exact-inference path; reducing it
  to a `JointGaussian` gives one for free.

---

## Philosophy — what a `JointGaussian` *is* (in #235's Q&A style)

**What is a `JointGaussian`?** A `Distribution[T]` where `T` is a (possibly nested) `Record` of
numeric blocks. It is a Gaussian measure over that structured event, carrying the full
cross-covariance between blocks. One draw is a `Record{block: array}`. It is the canonical
object on which **exact Gaussian conditioning** is defined; higher-level joints (#284 chains,
GP posteriors, ensemble methods) either *delegate* to it or *fit* it.

**What does it own?** Every exact Gaussian conditioning identity. Nothing above it re-implements
the algebra; they construct or fit a `JointGaussian` and call `condition_on`.

**What is its event?** Described by an `EventTemplate` (#235) — named, numeric, possibly nested
blocks. Block names are the field keys of draws and the keys used to address conditioning
(`condition_on(joint, y=...)`).

**How is its covariance represented?** Through a `GaussianRepresentation` seam. The first
implementation is `DenseRepresentation`; structured/`LinOp`-backed forms are future work behind
the same interface.

---

### The op contract is preserved

`condition_on(obj, ...) -> Distribution`. For a `JointGaussian`, conditioning on the values of
some blocks returns the **conditional `JointGaussian`** over the remaining blocks — a
`Distribution`. ✓ Conditioning on *all* blocks is an error (nothing left), as today.

### Conditioning = pushforward through an affine map

The result is *defined* as the pushforward of the joint through a conditional `AffineGaussianMap`.
By the Matheron lemma, if `(x_u, x_o) ~` joint prior, then

```
x_u + K (o* − x_o)  ~  p(x_u | x_o = o*),     K = Σ_uo Σ_oo⁻¹   (the gain)
```

This is an affine map *linear in the joint value*:

```
T(x_u, x_o) = [I, −K] · [x_u; x_o] + K · o*       # A = [I, −K],  b = K·o*
```

Pushing a Gaussian through `T` gives, in closed form,

```
mean:  m_u + K (o* − m_o)
cov:   [I, −K] Σ [I, −K]ᵀ  =  Σ_uu − Σ_uo Σ_oo⁻¹ Σ_ou
```

— exactly the conditioning identities. **The same `T` pushes samples** (the Matheron / ensemble
update), so one map yields both the parametric and the empirical conditional. Concretely:

```
condition_on(joint, o=o*)  ≡  pushforward(joint, joint.conditional_map("o"), value=o*)
```

and the pushforward registry (#67) carries a closed-form rule
`(AffineGaussianMap, JointGaussian) → JointGaussian`, so the op returns a `JointGaussian`
eagerly. (The same map applied to a `JointEmpirical` is the ensemble path — that rule lives in
the sibling issue.)

### Consistency with the composition algebra (#284)

#284 routes `condition_on` by field role:

- *exogenous given-field* → curry;
- *produced field, upstream / independent* → **exact slice** (forward-sampleable);
- *produced field, downstream data* (free ancestors) → Bayesian inversion → inference registry.

A `JointGaussian` is unconditional (no givens) and has **no ancestral DAG** — it is a symmetric
joint where every block is jointly Gaussian with every other. Conditioning any block on any other
is therefore **always exact and always forward-sampleable**: the conditional is itself a
`JointGaussian`. So `JointGaussian` only ever takes #284's *exact* path; it is precisely the
object that defines that path. A Gaussian `ChainDistribution`'s exact conditioning can be
implemented by reducing to a `JointGaussian` and conditioning here — that is the intended
consistency story (one place owns the algebra).

### Measurement noise / conditioning on a distribution

Conditioning on a *noisy* observation `o* = x_o + ε`, `ε ~ N(0, R)`, is "soft" Gaussian
conditioning: replace `Σ_oo` with the innovation covariance `S = Σ_oo + R` and `K = Σ_uo S⁻¹`.
This is the bridge to the ensemble-Kalman issue (where `R` is the observation-error covariance).
Two surface options for expressing it (open question below): an `obs_cov=R` kwarg on
`condition_on`, **or** modeling the noise as an explicit block (`from_generative`) so plain
conditioning suffices. For pure exact block conditioning, `R = 0` and the clean derivation above
holds.

---

## The `AffineGaussianMap` primitive

Lives in the `maps/` package (#67), as a `TransportMap` specialization; realizes #26's
`MatheronMap(AffineMap)`.

```python
cmap = joint.conditional_map(*observed_blocks, obs_cov=None, formula="auto") -> AffineGaussianMap
```

`conditional_map` builds the value-independent pieces — the gain `K` and the posterior
covariance `Σ_{u|o}` (these depend only on *which* blocks are observed, `R`, and `formula`, not
on the observed value). The observed value enters only the affine offset, so it is supplied when
the map is applied:

```python
cmap.push_moments(o_value)        -> JointGaussian        # parametric conditional
cmap.forward(joint_sample, o_value) -> Record             # deterministic Matheron core on one joint draw
cmap.pushforward(joint_or_empirical, o_value) -> Distribution   # registry dispatch (#67)
```

**Why split `K`/`Σ_post` (value-free) from the offset (value-dependent)?** The ensemble path
computes `K` once from moments and applies it with many different (perturbed) observation values
— so the reusable artifact is the value-free core. This split is what lets the sibling issue
build stochastic and square-root ensemble updates on top without recomputing the gain.

**L1/L2 boundary (stated here, used there).** The map this object produces is the
*deterministic affine core*. Stochastic (perturbed-obs) ensemble updates are the *same* core fed
perturbed observation values; deterministic square-root updates are a *different* construction
that reuses `K` for the mean and adds an anomaly transform — and act on the whole ensemble at
once. So `AffineGaussianMap` must support a **batched/ensemble action**, not only iid
member-wise. Keeping the stochasticization in the sibling issue keeps this object pure
deterministic moment algebra.

---

## The conditioning math — one identity, multiple computational forms

Partition into unobserved `u` (dim `p`) and observed `o` (dim `q`); observed value `o*`; optional
measurement noise `R`; innovation `S = Σ_oo + R`; gain `K = Σ_uo S⁻¹`.

```
m_{u|o} = m_u + K (o* − m_o)
Σ_{u|o} = Σ_uu − K Σ_ou
```

| form | solves in | best when | status |
|---|---|---|---|
| **observed-space** (gain) | `q` | few observed dims | **dense MVP** (today's path, + `R`) |
| **parameter/precision-space** | `p` | `p ≪ q`, cheap `R⁻¹` | **dense MVP** (`Σ_{u|o}⁻¹ = Σ_uu⁻¹ + HᵀR⁻¹H`) |
| **ensemble-space** | `J` (members) | `J ≪ p, q` (EnKF regime) | **future** (needs root/low-rank rep) |

`formula="auto"` picks by `min(p, q[, J])` and which representation/structure is available. The
two dense forms are mathematically equal; ensemble-space exploits a low-rank anomaly root and is
deferred to the structured representation (where `RootLinOp` + #31's `SolveLinOp`/`A⁻¹B` pay off).

All forms are pure `jnp` with no data-dependent shapes ⇒ **`jax.vmap`-clean**. Conditioning a
batch of `JointGaussian`s (per-cell spatial models, a `DistributionBatch`) vectorizes for free —
an invariant the implementation should preserve.

---

## The representation seam (dense now, `LinOp`-backed later)

`JointGaussian` keeps block names; the representation works in flat index space (mirrors today's
`_component_slices` + `jnp.ix_`, generalized through `EventTemplate.to_vector`). The interface
exposes exactly what conditioning needs — block accessors + solves — and nothing more:

```python
class GaussianRepresentation(ABC):
    @property
    def dim(self) -> int: ...
    def mean(self) -> Array                                  # flat (d,)
    def cov(self) -> LinOp                                    # full Σ (DenseLinOp now; lazy later)
    def select(self, idx: Array) -> GaussianRepresentation   # marginal over a block subset

    # Conditioning hook: default = generic algorithm driven by cov() blocks + LinOp.solve.
    # Structured reps override for an efficient closed form (e.g. ensemble-space).
    def condition(self, u_idx, o_idx, *, obs_cov=None, formula="auto") -> ConditionalCore: ...
```

- **`DenseRepresentation`** — first implementation; `cov()` is a `DenseLinOp`. Always correct.
- **`RootRepresentation`** (future) — `Σ = (1/(J−1)) A Aᵀ` with anomaly `LinOp` `A`; unlocks
  ensemble-space conditioning.
- **`CanonicalRepresentation`** (future) — precision/potential `(Λ, η)`; makes *combining*
  Gaussian factors cheap (a product of Gaussians ≈ `*` over Gaussian kernels, #284) and is the
  natural form for sequential assimilation of conditionally-independent observations.

Designing the seam now (even with only `DenseRepresentation` shipped) is the "dense default,
structured-ready" decision: the conditioning algorithm and public API are written against the
interface, so the future forms are additive.

---

### Construction

```python
JointGaussian(mean=…, cov=…, **block_shapes)                  # dense moment form (today; preserved)
JointGaussian.from_blocks(means={…}, covs={…}, cross={…})     # block-wise; builds a representation
JointGaussian.from_generative(prior=…, H=LinOp, noise=R)      # y = Hx + ε; implied cross/obs blocks
JointGaussian.from_root(mean, root=LinOp, **block_shapes)     # low-rank anomaly form (future rep)
```

`from_generative` (known `H`) and a moment-fit converter (`from_distribution(empirical,
JointGaussian)`, in the sibling issue) produce the *same* `JointGaussian{x, y}`, after which
conditioning is byte-identical — the exact-vs-ensemble unification made concrete.

### Operations

```python
condition_on(joint, **observed)        -> JointGaussian        # op; pushforward through conditional_map
joint.conditional_map(*blocks, obs_cov=None, formula="auto") -> AffineGaussianMap   # reusable primitive
joint.marginal(*blocks)                -> JointGaussian        # drop blocks (explicit marginalization)
sample / log_prob / mean / variance / cov                     # as today
```

### Properties

`fields`, `event_template`, `event_shapes`, `mean_vector`, `covariance`, `components`
(per-block marginals) — as today, extended for nesting.

---

## Marginalize vs. condition (naming fix)

Make marginalization explicit and distinct from value-conditioning:

- `joint.marginal(*blocks)` — return the Gaussian over `blocks` (drop the rest's rows/cols). Pure
  structure, no observed value.
- `condition_on(joint, block=value)` — value-aware conditioning (the affine-map pushforward).

This also flags the `JointEmpirical` wart: its current `_condition_on` is marginalization and
should be renamed/redirected so that `condition_on` on an empirical joint is the *value-aware*
(ensemble-Kalman) operation defined in the sibling issue.

---

## Relationship to existing code & issues

- **#235** — `EventTemplate` is the block schema; `to_vector`/`from_vector`/`numeric_subset`
  replace the hand-rolled flat slices and enable nested blocks.
- **#284** — `condition_on` consistency (above); a Gaussian `ChainDistribution` reduces here for
  exact conditioning.
- **#255** — kernels: an observation model is a `ConditionalDistribution`; pairing it with a
  Gaussian prior and reducing to a `JointGaussian` is the exact linear-Gaussian path.
- **#26 / #67** — `AffineGaussianMap` is a `TransportMap`; conditioning is registered as a
  closed-form pushforward. This proposal is essentially the unfinished half of #26's `MatheronMap`.
- **#31** — the future `RootRepresentation`/ensemble-space form needs `SolveLinOp`/`A⁻¹B` and
  exploits the recently-fixed `RootLinOp`/`CholeskyLinOp`.
- **`GaussianRandomFunction`** — GP regression could share this conditioning core.

---

## Migration / back-compat

- Today's `JointGaussian(mean=…, cov=…, **shapes)` constructor and dense `condition_on` behavior
  are preserved (the dense observed-space form is exactly current behavior with `R = 0`).
- New surface is additive: `conditional_map`, `marginal`, `from_*` constructors, `formula=`,
  `obs_cov=`.
- Internal: route the existing `_condition_on_impl` through `DenseRepresentation` +
  `AffineGaussianMap` so the math lives in one place.

---

## Open questions

1. **Measurement noise surface.** `obs_cov=R` kwarg on `condition_on` vs. modeling noise as an
   explicit block (`from_generative`)? The former is terser; the latter keeps `condition_on`
   purely about exact block conditioning and pushes noise into model construction.
2. **What the `AffineGaussianMap` stores.** Only `K` + posterior `Σ_uu`, or also the `q`-dim
   moments? Affects serialization and the batched ensemble action.
3. **`from_generative` home.** A `JointGaussian` classmethod, or should a linear-Gaussian
   `ChainDistribution` (#284) expose `as_joint_gaussian()` as the canonical bridge?
4. **Eager vs. lazy pushforward.** Should `condition_on` eagerly materialize the conditional
   `JointGaussian` (proposed), or return a lazy `PushforwardDistribution` that computes moments on
   demand? Eager is simpler and matches current behavior; lazy composes better with non-closed-form
   maps.
5. **`marginal` vs `select` naming**, and whether marginalization is a representation method, a
   `JointGaussian` method, or both.

**Comment by jhuggins** [comment]
Closing: absorbed into the design reference in revised form. The Gaussian algebra (`design/06-distribution-catalog.md`) fixes `FactoredMultivariateGaussian` as *derived, never constructed* — `*` / `joint` return it as the most-specific class whenever every factor is Gaussian or linear-Gaussian — with exact conditioning, marginals, and affine pushforward rules registered with the pushforward registry (#26, PR #67). This proposal's conditioning-as-pushforward idea survives as those closed-form rules. Implementation is tracked in #342; the ensemble-Kalman follow-on stays open as #299.
